### PR TITLE
Generic read method

### PR DESCRIPTION
**Pull request by bhazelton**

### CODE REVIEW

**Comment by bhazelton** [comment]
@plaplant can you review this PR?

**Comment by plaplant** [comment]
Sure, I'll take a look in the next ~2-3 days, if that's okay

**Comment by bhazelton** [comment]
yep, no big rush.

File: pyuvdata/tests/test_uvdata.py [hunk 1]
```diff
@@ -322,6 +322,24 @@ def test_HERA_diameters():
     uv_in.check()
 
 
+def test_generic_read():
+    uv_in = UVData()
+    uvfits_file = os.path.join(DATA_PATH, 'day2_TDEM0003_10s_norx_1src_1spw.uvfits')
+    uvh_file = os.path.join(DATA_PATH, 'test', 'outtest_uvfits.h5')
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/tests/test_uvdata.py]: No, that's cruft. Thanks for catching it!

File: pyuvdata/tests/test_uvfits.py [hunk 2]
```diff
@@ -67,14 +67,14 @@ def test_noSPW():
 #    """Test reading in an RTS UVFITS file."""
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/tests/test_uvfits.py]: Yes. It's very old, back when we were still validating our uvfits reader.

File: pyuvdata/uvdata.py [hunk 3]
```diff
@@ -2124,6 +2127,271 @@ def write_uvh5_part(self, filename, data_array, flags_array, nsample_array, chec
...
+             correct_lat_lon=True, use_model=False, data_column='DATA',
+             pol_order='AIPS', run_check=True, check_extra=True,
+             run_check_acceptability=True):
+        """
+        Read a generic file into a UVData object.
+
+        Args:
+            filename: The file to read.
+            file_type: One of ['uvfits', 'miriad', 'fhd', 'ms', 'uvh5'] or None.
+                If None, guess based on file extension.
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvdata.py]: yep, I'll make it clearer.

File: pyuvdata/uvdata.py [hunk 4]
```diff
@@ -2124,6 +2127,271 @@ def write_uvh5_part(self, filename, data_array, flags_array, nsample_array, chec
...
+                if extension == '.sav' or extension == '.txt':
+                    file_type = 'fhd'
+                    multi = False
+                else:
+                    multi = True
+        else:
+            multi = False
+
+        if file_type is None:
+            if multi:
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvdata.py]: sure.

File: pyuvdata/uvdata.py [hunk 5]
```diff
@@ -2124,6 +2127,271 @@ def write_uvh5_part(self, filename, data_array, flags_array, nsample_array, chec
...
+             times=None, polarizations=None, blt_inds=None, time_range=None,
+             read_metadata=True, read_data=True, phase_type=None,
+             correct_lat_lon=True, use_model=False, data_column='DATA',
+             pol_order='AIPS', run_check=True, check_extra=True,
+             run_check_acceptability=True):
+        """
+        Read a generic file into a UVData object.
+
+        Args:
+            filename: The file to read.
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvdata.py]: yes, thanks

File: pyuvdata/uvdata.py [hunk 6]
```diff
@@ -2124,6 +2127,271 @@ def write_uvh5_part(self, filename, data_array, flags_array, nsample_array, chec
...
+
+        if file_type is None:
+            raise ValueError('File type could not be determined.')
+
+        if (time_range is not None):
+            if times is not None:
+                raise ValueError(
+                    'Only one of times and time_range can be provided.')
+            if file_type == 'uvfits' or file_type == 'uvh5':
+                unique_times = np.unique(self.time_array)
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvdata.py]: you're right. I had it after the reads before which was more correct. I was trying to not duplicate code so I moved it up, but that doesn't always work. I'll move it back to after the read.

**Comment by bhazelton** [comment]
@plaplant I think I fixed everything, thanks for your suggestions!

File: pyuvdata/uvdata.py [hunk 7]
```diff
@@ -2124,6 +2127,277 @@ def write_uvh5_part(self, filename, data_array, flags_array, nsample_array, chec
...
+                    # It's miriad.
+                    file_type = 'miriad'
+                elif os.path.exists(os.path.join(file_test, 'OBSERVATION')):
+                    # It's a measurement set.
+                    file_type = 'ms'
+            else:
+                basename, extension = os.path.splitext(file_test)
+                if extension == '.uvfits':
+                    file_type = 'uvfits'
+                elif extension == '.h5':
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvdata.py]: oops, yes, let's be consistent. I'll let you make the changes if you don't mind.

**Pull request merged**